persist/fs/commitlog/reader.go

```diff
@@ -231,99 +210,122 @@ func (r *reader) startBackgroundWorkers() error {
-			r.decoderBufs[index%r.numConc] <- decoderArg{
-				bytes: data,
-				err:   err,
+			if err != nil {
+				if err == io.EOF {
+					return
+				}
+				r.decoderBufs[0] <- decoderArg{
+					bytes: data,
+					err:   err,
+				}
+				continue
 			}
-			index++
-			if err == io.EOF {
-				return
+
+			decoderStream.Reset(data)
+			decoder.Reset(decoderStream)
+			decodeRemainingToken, uniqueIndex, err := decoder.DecodeLogEntryUniqueIndex()
+
+			// Distribute work by the uniqueIndex so that each decoder loop is receiving
+			// all datapoints for a given series within relative order.
+			r.decoderBufs[uniqueIndex%uint64(r.numConc)] <- decoderArg{
+				bytes:                data,
+				err:                  err,
+				decodeRemainingToken: decodeRemainingToken,
+				uniqueIndex:          uniqueIndex,
+				offset:               decoderStream.Offset(),
 			}
 		}
 	}
 }
 
-func (r *reader) shutdown() {
-	for _, decoderBuf := range r.decoderBufs {
-		close(decoderBuf)
-	}
-	r.shutdownCh <- r.close()
-}
-
 func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- readResponse) {
 	var (
 		decodingOpts          = r.opts.FilesystemOptions().DecodingOptions()
 		decoder               = msgpack.NewDecoder(decodingOpts)
 		decoderStream         = msgpack.NewDecoderStream(nil)
 		metadataDecoder       = msgpack.NewDecoder(decodingOpts)
 		metadataDecoderStream = msgpack.NewDecoderStream(nil)
+		metadataLookup        = make(map[uint64]seriesMetadata)
 	)
+
 	for arg := range inBuf {
 		readResponse := readResponse{}
 		// If there is a pre-existing error, just pipe it through
 		if arg.err != nil {
 			readResponse.resultErr = arg.err
-			r.writeToOutBuf(outBuf, readResponse)
+			outBuf <- readResponse
 			continue
 		}
 
 		// Decode the log entry
-		arg.bytes.IncRef()
-		decoderStream.Reset(arg.bytes.Get())
+		decoderStream.Reset(arg.bytes[arg.offset:])
 		decoder.Reset(decoderStream)
-		entry, err := decoder.DecodeLogEntry()
+		entry, err := decoder.DecodeLogEntryRemaining(arg.decodeRemainingToken, arg.uniqueIndex)
 		if err != nil {
 			readResponse.resultErr = err
-			r.writeToOutBuf(outBuf, readResponse)
+			outBuf <- readResponse
 			continue
 		}
 
 		// If the log entry has associated metadata, decode that as well
 		if len(entry.Metadata) != 0 {
-			err := r.decodeAndHandleMetadata(metadataDecoder, metadataDecoderStream, entry)
+			err := r.decodeAndHandleMetadata(metadataLookup, metadataDecoder, metadataDecoderStream, entry)
 			if err != nil {
 				readResponse.resultErr = err
-				r.writeToOutBuf(outBuf, readResponse)
+				outBuf <- readResponse
 				continue
 			}
 		}
 
-		// Metadata may or may not actually be nonzero here, the consumer in Read()
-		// will wait if so
-		readResponse.series, readResponse.pendingMetadata = r.lookupMetadata(entry.Index)
+		metadata, hasMetadata := metadataLookup[entry.Index]
+		if !hasMetadata {
+			// Corrupt commit log
+			readResponse.resultErr = errCommitLogReaderMissingMetadata
```

Comment:
might be useful to include the series ID in the returned error. 

also, put a little note explaining how you know this is corrupt (the whole ordering guarantee your decoding scheme provides).

Reply:
good point

Reply:
actually I added the comment, but I can't include the ID because I don't have the metadata :P

Reply:
lol now that you say this, i'm pretty sure i asked the same thing the last time. mind putting a comment with this nuance here.